activerecord/test/cases/finder_test.rb

```diff
@@ -1810,15 +1810,17 @@ def test_finder_with_offset_string
   end
 
   test "#skip_query_cache! for #exists? with a limited eager load" do
-    Topic.cache do
-      assert_queries(1) do
-        Topic.eager_load(:replies).limit(1).exists?
-        Topic.eager_load(:replies).limit(1).exists?
-      end
+    Topic.connection.unprepared_statement do
```

Comment:
Hmm, what will happen if you are using prepared statements? It feels bad if these queries can't be cached in the query cache.

Reply:
For some unknown reason the prepared version of the query isn't cached. Note that setting `collector.preparable = false` like the existing `HomogeneousIn` implementation doesn't help.

Reply:
It looks like `ActiveModel::Attribute::WithCastValue`'s `value_before_type_cast`, which is used in its `hash`, will contain `ActiveRecord::ConnectionAdapters::PostgreSQL::OID::Array::Data` which doesn't have a consistent hash, which seems to be caused by the `PG::TextEncoder::Array` it contains.

Reply:
Providing a custom `hash` implementation seems to work. I'll commit this soon:
```rb
Data = Struct.new(:encoder, :values) do # :nodoc:
  def hash
    [encoder.name, encoder.delimiter, values].hash
  end
end